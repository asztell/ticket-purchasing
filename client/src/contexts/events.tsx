import { createContext, useEffect, useMemo, useState } from "react";

export type Event = {
  id: string;
  name: string;
  location: string;
  ISODate: string;
  price: number;
};

export type EventsContextType = {
  events: Event[];
  error: string | null;
};

export const EventsContext = createContext<EventsContextType>(
  {} as EventsContextType
);

export function EventsProvider({ children }: { children: React.ReactNode }) {
  const [events, setEvents] = useState<Event[]>([]);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    // const url = `${document.location.origin}/events`;
    const url = `${process.env.REACT_APP_API}/events`;
    console.log("url", url);
    fetch(url)
      .then((res) => {
        if (!res.ok) {
          throw res;
        }
        return res.json();
      })
      .then((result) => {
        console.log("events", result);
        setEvents(result);
      })
      .catch((error) => {
        setError(error);
      });
  }, [setEvents]);

  const value = useMemo(() => {
    return {
      events,
      error,
    };
  }, [events, error]);

  return (
    <EventsContext.Provider value={value}>{children}</EventsContext.Provider>
  );
}
