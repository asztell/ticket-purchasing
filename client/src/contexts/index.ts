export * from './events'
export * from './ticketPurchasing'
