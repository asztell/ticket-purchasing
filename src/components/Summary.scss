.Total {
  border: 1px solid rgb(191, 191, 191);
  border-radius: 4px;
  padding-left: 20px;
  padding-right: 20px;
  & > button {
    border-radius: 2px;
    border: 0 solid transparent;
    height: 39px;
    padding: 10px 25%;
    background-color: rgb(20 128 22);
    color: #ffffff;
    font-weight: bold;
    margin-top: 25px;
    margin-bottom: 25px;
    width: 100%;
    &:disabled {
      background-color: rgb(2, 108, 223);
      opacity: 0.2;
    }
  }
  & > h2 {
    font-size: 20px;
    font-family: "Averta", "Helvetica Neue", Helvetica, Arial, sans-serif;
  }
}
