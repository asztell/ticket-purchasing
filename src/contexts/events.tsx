import { createContext, useEffect, useMemo, useState } from "react";

export type Event = {
  id: string;
  name: string;
  location: string;
  ISODate: string;
  price: number;
};

export type EventsContextType = {
  events: Event[];
  error: string | null;
};

export const EventsContext = createContext<EventsContextType>(
  {} as EventsContextType
);

export function EventsProvider({ children }: { children: React.ReactNode }) {
  const [events, setEvents] = useState<Event[]>([]);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    fetch(`http://localhost:8080/events`)
      .then((res) => {
        if (!res.ok) {
          throw res;
        }
        return res.json();
      })
      .then((result) => {
        setEvents(result);
      })
      .catch((error) => {
        setError(error);
      });
  }, [setEvents]);

  const value = useMemo(() => {
    return {
      events,
      error,
    };
  }, [events, error]);

  return (
    <EventsContext.Provider value={value}>{children}</EventsContext.Provider>
  );
}
