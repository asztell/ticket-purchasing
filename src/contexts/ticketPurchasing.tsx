import { createContext, useCallback, useMemo, useState } from "react";
import { Event } from "./events";

export type CardInfoType = {
  nameOnCard: string;
  cardNumber: string;
  cardType: string;
  securityCode: string;
  securityCodeValid: boolean;
  expirationDate: string;
  expirationDateValid: boolean;
};

export type TicketPurchasingContextType = {
  selectedEvent: Event | undefined;
  ticketsCounter: number;
  cardInfo: CardInfoType;
  updateSelectedEvent: (selectedEvent: Event) => void;
  updateTicketsCounter: (ticketsCounter: number) => void;
  updateCardInfo: (cardInfo: CardInfoType) => void;
};

export const TicketPurchasingContext =
  createContext<TicketPurchasingContextType>({} as TicketPurchasingContextType);

export function TicketPurchasingProvider({
  children,
}: {
  children: React.ReactNode;
}) {
  const [ticketsCounter, setTicketsCounter] = useState(0);
  const [selectedEvent, setSelectedEvent] = useState<Event | undefined>(
    undefined
  );
  const [cardInfo, setCardInfo] = useState({
    nameOnCard: "",
    cardNumber: "",
    cardType: "",
    securityCode: "",
    securityCodeValid: false,
    expirationDate: "",
    expirationDateValid: false,
  });

  const updateTicketsCounter = useCallback((ticketsCounter: number) => {
    setTicketsCounter(
      (prevTicketsCounter) => prevTicketsCounter + ticketsCounter
    );
  }, []);

  const updateSelectedEvent = useCallback((selectedEvent: Event) => {
    setSelectedEvent(selectedEvent);
  }, []);

  const updateCardInfo = useCallback((cardInfo: CardInfoType) => {
    const { expirationDate, cardNumber, securityCode } = cardInfo;
    const cardType = validateCardNumber(cardNumber);
    const securityCodeValid =
      securityCode.match(new RegExp("^[0-9]{3,4}$")) !== null;
    const expirationDateValid = validateExpirationDate(expirationDate);
    setCardInfo({
      ...cardInfo,
      cardType,
      securityCodeValid,
      expirationDateValid,
    });
  }, []);

  const value = useMemo(() => {
    return {
      selectedEvent,
      ticketsCounter,
      cardInfo,
      updateSelectedEvent,
      updateTicketsCounter,
      updateCardInfo,
    };
  }, [
    selectedEvent,
    ticketsCounter,
    cardInfo,
    updateSelectedEvent,
    updateTicketsCounter,
    updateCardInfo,
  ]);
  return (
    <TicketPurchasingContext.Provider value={value}>
      {children}
    </TicketPurchasingContext.Provider>
  );
}

export function validateCardNumber(cardNumber: string) {
  const visa = new RegExp("^4[0-9]{12}(?:[0-9]{3})?$");
  if (cardNumber.match(visa) !== null) return "Visa";
  const mastercard = new RegExp(
    "^5[1-5][0-9]{14}|^(222[1-9]|22[3-9]\\d|2[3-6]\\d{2}|27[0-1]\\d|2720)[0-9]{12}$"
  );
  if (cardNumber.match(mastercard) !== null) return "Mastercard";
  const amex = new RegExp("^3[47][0-9]{13}$");
  if (cardNumber.match(amex)) return "American Express";
  const discover = new RegExp("^6(?:011|5[0-9]{2})[0-9]{12}$");
  if (cardNumber.match(discover)) return "Discover";
  const dinersClub = new RegExp("^3(?:0[0-5]|[68][0-9])[0-9]{11}$");
  if (cardNumber.match(dinersClub)) return "Diners Club";
  if (cardNumber.length === 0) return "";
  return "Invalid";
}

export function validateExpirationDate(expirationDate: string) {
  const expirationDateValid =
    expirationDate.match(
      new RegExp("^(0[1-9]|1[0-2])\\/?([0-9]{4}|[0-9]{2})$")
    ) !== null && new Date(expirationDate) > new Date();
  return expirationDateValid;
}
