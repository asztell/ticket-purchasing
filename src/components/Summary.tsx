import { useContext, useCallback } from "react";
import { useNavigate } from "react-router-dom";
import { TicketPurchasingContext } from "../contexts";
import "./Summary.scss";

export function Summary({ className }: { className?: string }) {
  const {
    selectedEvent,
    ticketsCounter,
    cardInfo,
    termsOfUseChecked,
    updateTermsOfUseChecked,
  } = useContext(TicketPurchasingContext);
  const {
    cardType,
    securityCodeValid,
    expirationDateValid,
    cardNumber,
    securityCode,
  } = cardInfo;

  const handleChangeTermsOfUse = useCallback(() => {
    updateTermsOfUseChecked(!termsOfUseChecked);
  }, [termsOfUseChecked, updateTermsOfUseChecked]);

  const navigate = useNavigate();
  const handleSubmit = useCallback(async () => {
    console.log("Submitting...");
    // TODO: start spinner (loading)
    try {
      // const url = `${document.location.origin}/checkout`;
      const url = `${process.env.REACT_APP_API}/checkout`;
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          event: selectedEvent,
          tickets: ticketsCounter,
          cardNumber: cardNumber,
          securityCode: securityCode,
        }),
      });
      console.log("/checkout response", response);
      if (!response.ok) {
        throw response;
      }
      const json = await response.json();
      navigate("/confirmation", { state: json });
    } catch (error) {
      navigate("/confirmation", { state: error });
      console.log(error);
    }
    // TODO: stop spinner/loading (in case user comes back to this page)
  }, [selectedEvent, ticketsCounter, navigate, cardNumber, securityCode]);

  return (
    <div className={className}>
      <div className="Total">
        <h2>Total</h2>
        <p>Event: {selectedEvent?.name}</p>
        <p>Tickets: {ticketsCounter}</p>
        <p>Card Number: {cardNumber}</p>
        <p>Security Code: {securityCode}</p>
        <input
          type="checkbox"
          onChange={handleChangeTermsOfUse}
          checked={termsOfUseChecked}
          name="termsOfUse"
        />
        <label htmlFor="termsOfUse">
          I have read and agree to the current{" "}
          <a href="https://google.com">Terms of Use</a>
        </label>
        <button
          className="Purchase-Tickets-Button"
          onClick={handleSubmit}
          disabled={
            cardType === "" ||
            cardType === "Invalid" ||
            !securityCodeValid ||
            !expirationDateValid ||
            !termsOfUseChecked ||
            ticketsCounter === 0
          }
        >
          Place Order
        </button>
      </div>
    </div>
  );
}
