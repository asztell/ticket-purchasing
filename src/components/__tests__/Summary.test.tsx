import { render } from "@testing-library/react";
import { Summary } from "../Summary";
import { TicketPurchasingProvider } from "../../contexts";
import { MemoryRouter as Router } from "react-router-dom";

describe("<Summary />", () => {
  test("render Summary", () => {
    const { container } = render(
      <Router>
        <TicketPurchasingProvider>
          <Summary />
        </TicketPurchasingProvider>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });
});
