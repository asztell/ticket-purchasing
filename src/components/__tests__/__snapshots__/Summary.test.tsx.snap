// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Summary /> render Summary 1`] = `
<div>
  <div>
    <div
      class="Total"
    >
      <h2>
        Total
      </h2>
      <p>
        Event: 
      </p>
      <p>
        Tickets: 
        0
      </p>
      <p>
        Card Number: 
      </p>
      <p>
        Security Code: 
      </p>
      <input
        name="termsOfUse"
        type="checkbox"
      />
      <label
        for="termsOfUse"
      >
        I have read and agree to the current 
        <a
          href="#"
        >
          Terms of Use
        </a>
      </label>
      <button
        class="Purchase-Tickets-Button"
        disabled=""
      >
        Place Order
      </button>
    </div>
  </div>
</div>
`;
